### PR TITLE
Add CRD Conversion Strategy for forward compatibility

### PR DESCRIPTION
**Pull request by mbohlool**
Current implementation of CRD versioning assuming no-op conversion by default. Adding new conversion in future releases would have a downgrade problem as current API server would default back to no-op. see #65790 for discussion on this.

This PR is a possible solution to that, consider it as a proposal.

### CODE REVIEW

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types.go [hunk 1]
```diff
@@ -136,6 +152,9 @@ const (
 	NamesAccepted CustomResourceDefinitionConditionType = "NamesAccepted"
 	// Terminating means that the CustomResourceDefinition has been deleted and is cleaning up.
 	Terminating CustomResourceDefinitionConditionType = "Terminating"
+	// Valid means the CRD does pass validation. K8S objects are not validated on the GET path and it is
+	// possible to have invalid objects on downgrade/upgrade paths.
+	Valid CustomResourceDefinitionConditionType = "Valid"
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types.go]: Isn't validation an invariant of every object in etcd we always try to guarantee? How can it become invalid?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation_test.go [hunk 2]
```diff
@@ -121,6 +161,9 @@ func TestValidateCustomResourceDefinition(t *testing.T) {
 							Storage: true,
 						},
 					},
+					Conversion: &apiextensions.CustomResourceConversion{
+						Strategy: apiextensions.ConversionStrategyType("nop"),
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation_test.go]: nop or no-op?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go [hunk 3]
```diff
@@ -26,21 +26,31 @@ import (
...
-	// The only converter right now is nopConverter. More converters will be returned based on the
-	// CRD object when they introduced.
-	unsafe = &crdConverter{
-		clusterScoped: crd.Spec.Scope == apiextensions.ClusterScoped,
-		delegate: &nopConverter{
-			validVersions: validVersions,
-		},
+	// if crd.Spec.Conversion is nil, that means the CRD created on an API server that do not have this field yet.
+	// That means we can default strategy to NopConverter.
+	Strategy := apiextensions.NopConverter
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go]: Elsewhere we default the value in that case. I.e. in the registry code we will never see an unset field.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go [hunk 4]
```diff
@@ -26,21 +26,31 @@ import (
...
+		unsafe = &crdConverter{
+			clusterScoped: crd.Spec.Scope == apiextensions.ClusterScoped,
+			delegate: &nopConverter{
+				validVersions: validVersions,
+			},
+		}
+		return &safeConverterWrapper{unsafe}, unsafe, nil
+	}
+
+	return nil, nil, fmt.Errorf("unknown conversion strategy \"%s\" for CRD %s", crd.Spec.Conversion.Strategy, crd.Spec.Group)
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go]: `crd.Spec.Group` is only the group. The resource name is missing.

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/testserver/start.go [hunk 5]
```diff
@@ -124,38 +132,51 @@ func StartServer(config *extensionsapiserver.Config) (chan struct{}, *rest.Confi
 	})
 	if err != nil {
 		close(stopCh)
-		return nil, nil, err
+		return nil, nil, nil, err
 	}
 
-	return stopCh, config.GenericConfig.LoopbackClientConfig, nil
+	customResourceDefinitionStorage := customresourcedefinition.NewREST(apiserver.Scheme, config.Complete().GenericConfig.RESTOptionsGetter)
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/test/integration/testserver/start.go]: re-use the completed config from above.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types.go (hunk 1 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types.go]: With @liggitt's comment https://github.com/kubernetes/kubernetes/issues/65790#issuecomment-404509214 in mind, I don't think this is a path worth pursuing. We have the very same problem of data-loss for any rollback which takes away features and fields that we used before rollback. Our cure: if you want to rollback, wait one Kube version iteration more until your rollback target version supports what you use.

Also note that all future conversion features will start as alpha with a feature gate. I.e. we have to plan in advance the transition from alpha to beta, and possibly a rollback to alpha. In that case, we can keep serving and understanding new validation features in a graceful way, even when rolling back to kube where the feature is still alpha.

Comment (mbohlool) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types.go]: I should remove this, was remnant of another solution.

> Also note that all future conversion features will start as alpha with a feature gate.

I do not disagree but still thinking about this :)

Generally I think it is still good to have this in 1.11.2. I commented on the related issue.

**Pull request closed without merging**